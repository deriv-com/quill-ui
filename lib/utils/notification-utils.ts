export const TYPE = {
    ERROR: "error",
    INFO: "info",
    SUCCESS: "success",
    WARNING: "warning",
} as const;

export const STATUS = {
    READ: "read",
    UNREAD: "unread",
};
