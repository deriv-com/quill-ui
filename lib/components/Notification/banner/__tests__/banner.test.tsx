import { ComponentProps } from "react";
import { render, screen, waitFor } from "@testing-library/react";
import NotificationBanner from "..";
import { TYPE } from "@utils/notification-utils";
import userEvent from "@testing-library/user-event";

describe("NotificationBanner", () => {
    const customIcon = "Custom icon";
    const testMessage = "Test message";
    const testTitle = "Test title";
    const mokedProps: ComponentProps<typeof NotificationBanner> = {
        className: "test",
        message: testMessage,
        onClose: jest.fn(),
        onClick: jest.fn(),
        redirectTo: "https://www.example.com",
        title: testTitle,
        type: TYPE.INFO,
    };
    it("should display correct content", () => {
        const { container } = render(<NotificationBanner {...mokedProps} />);

        expect(container).toMatchSnapshot();
        expect(screen.getByText(testTitle)).toBeInTheDocument();
        expect(screen.getByText(testMessage)).toBeInTheDocument();
        expect(
            screen.getByRole("button", { name: "close" }),
        ).toBeInTheDocument();
    });
    it("should display correct banner with a custom icon", () => {
        const { container } = render(
            <NotificationBanner
                {...mokedProps}
                icon={<div>{customIcon}</div>}
                iconBackgroundColor="#000000"
            />,
        );

        expect(container).toMatchSnapshot();
        expect(screen.getByText(customIcon)).toBeInTheDocument();
    });
    it("should render correctly for mobile - without close button", () => {
        const { container } = render(
            <NotificationBanner {...mokedProps} isMobile />,
        );

        expect(container).toMatchSnapshot();
        expect(
            screen.queryByRole("button", { name: "close" }),
        ).not.toBeInTheDocument();
    });
    it("should render notification correctly based on the notification type", () => {
        const { container } = render(
            <NotificationBanner {...mokedProps} type={TYPE.WARNING} />,
        );

        expect(container).toMatchSnapshot();
    });
    it("should call onClick after banner is clicked", async () => {
        render(<NotificationBanner {...mokedProps} />);

        const notification = screen.getByText(testMessage);
        await userEvent.click(notification);

        expect(mokedProps.onClick).toHaveBeenCalled();
    });
    it("should call onClose after close button is clicked", async () => {
        render(<NotificationBanner {...mokedProps} />);

        const closeButton = screen.getByRole("button", { name: "close" });
        await userEvent.click(closeButton);

        await waitFor(() => {
            expect(mokedProps.onClose).toHaveBeenCalled();
        });
    });
    it("should hide after autohideTimeout + 240ms of animation", () => {
        jest.useFakeTimers();
        const { container } = render(
            <NotificationBanner {...mokedProps} autohideTimeout={4000} />,
        );

        jest.advanceTimersByTime(4240);
        expect(mokedProps.onClose).toHaveBeenCalled();
        jest.useRealTimers();
        expect(container).toMatchSnapshot();
    });
});
