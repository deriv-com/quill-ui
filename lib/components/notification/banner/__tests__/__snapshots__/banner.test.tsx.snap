// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBanner should apply correct classname 1`] = `
<div>
  <div
    class="notification-banner test"
  >
    <div
      class="body"
    >
      <div
        class="icon info"
      >
        <svg
          fill="#1789E1"
          height="24"
          role="img"
          viewBox="0 0 32 32"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 8.375c-2.93 0-5.586 1.563-7.07 4.063-1.446 2.539-1.446 5.624 0 8.124A8.13 8.13 0 0 0 16 24.626c2.89 0 5.547-1.523 7.031-4.062 1.446-2.5 1.446-5.586 0-8.125-1.484-2.5-4.14-4.063-7.031-4.063M16 26.5c-3.594 0-6.875-1.875-8.672-5-1.797-3.086-1.797-6.875 0-10 1.797-3.086 5.078-5 8.672-5 3.555 0 6.836 1.914 8.633 5 1.797 3.125 1.797 6.914 0 10a9.93 9.93 0 0 1-8.633 5m-1.562-6.875h.937v-2.5h-.937c-.547 0-.938-.39-.938-.937 0-.508.39-.938.938-.938h1.874a.95.95 0 0 1 .938.938v3.437h.313a.95.95 0 0 1 .937.938c0 .546-.43.937-.937.937h-3.125c-.547 0-.938-.39-.938-.937 0-.508.39-.938.938-.938M16 14c-.703 0-1.25-.547-1.25-1.25 0-.664.547-1.25 1.25-1.25.664 0 1.25.586 1.25 1.25 0 .703-.586 1.25-1.25 1.25"
          />
        </svg>
      </div>
      <div
        class="content"
      >
        <p
          class="quill-typography__body-text__size--sm__weight--bold__decoration--default quill-typography__color--default title"
        />
        <p
          class="quill-typography__caption-text__weight--regular__decoration--default quill-typography__color--prominent message"
        />
      </div>
    </div>
    <button
      class="icon close"
    >
      <svg
        height="22"
        role="img"
        viewBox="0 0 11 22"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M9.11 15.016 5.5 11.379l-3.637 3.637c-.164.164-.437.164-.629 0-.164-.192-.164-.465 0-.63l3.637-3.636-3.637-3.61c-.164-.163-.164-.437 0-.628.192-.164.465-.164.63 0L5.5 10.148l3.61-3.636c.163-.164.437-.164.628 0 .164.191.164.465 0 .629L6.102 10.75l3.636 3.637c.164.164.164.437 0 .629-.191.164-.465.164-.629 0"
          />
        </g>
        <defs>
          <clippath
            id="ff78866e2c6e70ea03c36ee4f3153a64__a"
          >
            <path
              d="M0 0h11v22H0z"
            />
          </clippath>
        </defs>
      </svg>
    </button>
  </div>
</div>
`;

exports[`NotificationBanner should have correct title 1`] = `
<div>
  <div
    class="notification-banner"
  >
    <div
      class="body"
    >
      <div
        class="icon info"
      >
        <svg
          fill="#1789E1"
          height="24"
          role="img"
          viewBox="0 0 32 32"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 8.375c-2.93 0-5.586 1.563-7.07 4.063-1.446 2.539-1.446 5.624 0 8.124A8.13 8.13 0 0 0 16 24.626c2.89 0 5.547-1.523 7.031-4.062 1.446-2.5 1.446-5.586 0-8.125-1.484-2.5-4.14-4.063-7.031-4.063M16 26.5c-3.594 0-6.875-1.875-8.672-5-1.797-3.086-1.797-6.875 0-10 1.797-3.086 5.078-5 8.672-5 3.555 0 6.836 1.914 8.633 5 1.797 3.125 1.797 6.914 0 10a9.93 9.93 0 0 1-8.633 5m-1.562-6.875h.937v-2.5h-.937c-.547 0-.938-.39-.938-.937 0-.508.39-.938.938-.938h1.874a.95.95 0 0 1 .938.938v3.437h.313a.95.95 0 0 1 .937.938c0 .546-.43.937-.937.937h-3.125c-.547 0-.938-.39-.938-.937 0-.508.39-.938.938-.938M16 14c-.703 0-1.25-.547-1.25-1.25 0-.664.547-1.25 1.25-1.25.664 0 1.25.586 1.25 1.25 0 .703-.586 1.25-1.25 1.25"
          />
        </svg>
      </div>
      <div
        class="content"
      >
        <p
          class="quill-typography__body-text__size--sm__weight--bold__decoration--default quill-typography__color--default title"
        >
          Test title
        </p>
        <p
          class="quill-typography__caption-text__weight--regular__decoration--default quill-typography__color--prominent message"
        />
      </div>
    </div>
    <button
      class="icon close"
    >
      <svg
        height="22"
        role="img"
        viewBox="0 0 11 22"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M9.11 15.016 5.5 11.379l-3.637 3.637c-.164.164-.437.164-.629 0-.164-.192-.164-.465 0-.63l3.637-3.636-3.637-3.61c-.164-.163-.164-.437 0-.628.192-.164.465-.164.63 0L5.5 10.148l3.61-3.636c.163-.164.437-.164.628 0 .164.191.164.465 0 .629L6.102 10.75l3.636 3.637c.164.164.164.437 0 .629-.191.164-.465.164-.629 0"
          />
        </g>
        <defs>
          <clippath
            id="ff78866e2c6e70ea03c36ee4f3153a64__a"
          >
            <path
              d="M0 0h11v22H0z"
            />
          </clippath>
        </defs>
      </svg>
    </button>
  </div>
</div>
`;

exports[`NotificationBanner should render correct icon based on the notification type 1`] = `
<div>
  <div
    class="notification-banner"
  >
    <div
      class="body"
    >
      <div
        class="icon warning"
      >
        <svg
          fill="#E18D00"
          height="24"
          role="img"
          viewBox="0 0 32 32"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.688 9.82 7.953 22.516c-.078.117-.078.195-.078.312 0 .313.234.547.547.547h15.117c.313 0 .586-.234.586-.547 0-.117-.04-.195-.117-.312L16.273 9.82A.29.29 0 0 0 16 9.625c-.156 0-.234.078-.312.195m-1.602-.976A2.25 2.25 0 0 1 16 7.75c.742 0 1.484.43 1.875 1.094l7.734 12.695c.235.39.391.82.391 1.29 0 1.327-1.094 2.421-2.46 2.421H8.421A2.434 2.434 0 0 1 6 22.828q0-.703.352-1.289zm3.164 12.031c0 .703-.586 1.25-1.25 1.25-.703 0-1.25-.547-1.25-1.25 0-.664.547-1.25 1.25-1.25.664 0 1.25.586 1.25 1.25m-.312-7.187v3.75c0 .546-.43.937-.938.937-.547 0-.937-.39-.937-.937v-3.75c0-.508.39-.938.937-.938a.95.95 0 0 1 .938.938"
          />
        </svg>
      </div>
      <div
        class="content"
      >
        <p
          class="quill-typography__body-text__size--sm__weight--bold__decoration--default quill-typography__color--default title"
        />
        <p
          class="quill-typography__caption-text__weight--regular__decoration--default quill-typography__color--prominent message"
        />
      </div>
    </div>
    <button
      class="icon close"
    >
      <svg
        height="22"
        role="img"
        viewBox="0 0 11 22"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M9.11 15.016 5.5 11.379l-3.637 3.637c-.164.164-.437.164-.629 0-.164-.192-.164-.465 0-.63l3.637-3.636-3.637-3.61c-.164-.163-.164-.437 0-.628.192-.164.465-.164.63 0L5.5 10.148l3.61-3.636c.163-.164.437-.164.628 0 .164.191.164.465 0 .629L6.102 10.75l3.636 3.637c.164.164.164.437 0 .629-.191.164-.465.164-.629 0"
          />
        </g>
        <defs>
          <clippath
            id="ff78866e2c6e70ea03c36ee4f3153a64__a"
          >
            <path
              d="M0 0h11v22H0z"
            />
          </clippath>
        </defs>
      </svg>
    </button>
  </div>
</div>
`;

exports[`NotificationBanner should render correctly for mobile - without close button 1`] = `
<div>
  <div
    class="notification-banner"
  >
    <div
      class="body"
    >
      <div
        class="icon info"
      >
        <svg
          fill="#1789E1"
          height="24"
          role="img"
          viewBox="0 0 32 32"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 8.375c-2.93 0-5.586 1.563-7.07 4.063-1.446 2.539-1.446 5.624 0 8.124A8.13 8.13 0 0 0 16 24.626c2.89 0 5.547-1.523 7.031-4.062 1.446-2.5 1.446-5.586 0-8.125-1.484-2.5-4.14-4.063-7.031-4.063M16 26.5c-3.594 0-6.875-1.875-8.672-5-1.797-3.086-1.797-6.875 0-10 1.797-3.086 5.078-5 8.672-5 3.555 0 6.836 1.914 8.633 5 1.797 3.125 1.797 6.914 0 10a9.93 9.93 0 0 1-8.633 5m-1.562-6.875h.937v-2.5h-.937c-.547 0-.938-.39-.938-.937 0-.508.39-.938.938-.938h1.874a.95.95 0 0 1 .938.938v3.437h.313a.95.95 0 0 1 .937.938c0 .546-.43.937-.937.937h-3.125c-.547 0-.938-.39-.938-.937 0-.508.39-.938.938-.938M16 14c-.703 0-1.25-.547-1.25-1.25 0-.664.547-1.25 1.25-1.25.664 0 1.25.586 1.25 1.25 0 .703-.586 1.25-1.25 1.25"
          />
        </svg>
      </div>
      <div
        class="content"
      >
        <p
          class="quill-typography__body-text__size--sm__weight--bold__decoration--default quill-typography__color--default title"
        />
        <p
          class="quill-typography__caption-text__weight--regular__decoration--default quill-typography__color--prominent message"
        />
      </div>
    </div>
  </div>
</div>
`;
